### PR TITLE
Adding keystone IdP

### PR DESCRIPTION
**Pull request by sseago**
This now works with Keystone v3 both with and without SSL.

Docs issue: https://github.com/openshift/openshift-docs/issues/1073

### CODE REVIEW

File: pkg/cmd/server/api/validation/oauth.go [hunk 1]
```diff
@@ -159,6 +162,22 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
...
 }
 
+func ValidateKeystoneIdentityProvider(provider *api.KeystonePasswordIdentityProvider, identityProvider api.IdentityProvider) ValidationResults {
+	validationResults := ValidationResults{}
+
+	if len(provider.RemoteConnectionInfo.URL) == 0 {
+		validationResults.AddErrors(fielderrors.NewFieldRequired("provider.url"))
+		return validationResults
+	}
+	//FIXME: we should probably allow http -- do we ignore, or do we warn here?
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: warn

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
+package keystonepassword
+
+import (
+	"fmt"
+	"errors"
+	"net/http"
+
+	"github.com/golang/glog"
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack"
```
Comment (liggitt) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: need to add the godeps for this... how big are those?

Comment (deads2k) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: > need to add the godeps for this... how big are those?

When you add them, see the commit format.  We keep the godeps in different commits starting with `bump(dependency name): commit_id`

File: pkg/cmd/server/api/validation/oauth.go [hunk 3]
```diff
@@ -159,6 +162,22 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
 	return validationResults
 }
 
+func ValidateKeystoneIdentityProvider(provider *api.KeystonePasswordIdentityProvider, identityProvider api.IdentityProvider) ValidationResults {
+	validationResults := ValidationResults{}
+
+	if len(provider.RemoteConnectionInfo.URL) == 0 {
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: should be able to just call ValidateRemoteConnectionInfo, which will check the url, ca, and client cert stuff

Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: then warn about non-https separately

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go (hunk 2 above)
Comment (sseago) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: Ahh ok. These are already referenced in upstream kubernetes -- I wasn't sure if direct vs. indirect deps made a difference.

Comment (sseago) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: @liggitt @deads2k Looks like the gophercloud deps are already here as of this commit: https://github.com/openshift/origin/commit/83f9bfd2d4e5ecabfbd46c67f78edda058656051

**Comment by sseago** [comment]
Updated the commit:
1) incorporated feedback
2) SSL working
3) moved the hard-coded DomainName into a config param.

Unit testing is still missing.

**Comment by sseago** [comment]
Added panic recovery

File: pkg/cmd/server/api/validation/oauth.go [hunk 4]
```diff
@@ -134,6 +134,10 @@ func ValidateIdentityProvider(identityProvider api.IdentityProvider) ValidationR
 		case (*api.LDAPPasswordIdentityProvider):
 			validationResults.Append(ValidateLDAPIdentityProvider(provider))
 
+		case (*api.KeystonePasswordIdentityProvider):
+			validationResults.AddErrors(ValidateRemoteConnectionInfo(provider.RemoteConnectionInfo).Prefix("provider")...)
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: move the ValidateRemoteConnectionInfo call into ValidateKeystoneIdentityProvider

File: pkg/cmd/server/api/validation/oauth.go [hunk 5]
```diff
@@ -160,6 +164,18 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
 	return validationResults
 }
 
+// RemoteConnection fields validated separately -- this is for keystone-specific validation
+func ValidateKeystoneIdentityProvider(provider *api.KeystonePasswordIdentityProvider, identityProvider api.IdentityProvider) ValidationResults {
+	validationResults := ValidationResults{}
+
+	if !strings.HasPrefix(provider.RemoteConnectionInfo.URL, "https") {
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: parse the url and check the scheme, rather than prefix

File: pkg/cmd/server/api/validation/oauth.go [hunk 6]
```diff
@@ -160,6 +164,18 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
...
+		validationResults.AddWarnings(fielderrors.NewFieldInvalid("provider.url", provider.RemoteConnectionInfo.URL, "Auth URL should be secure and start with https"))
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: `NewFieldInvalid("url", ...)`, and let the caller prefix with "provider"

File: pkg/cmd/server/origin/auth.go [hunk 7]
```diff
@@ -520,6 +522,21 @@ func (c *AuthConfig) getPasswordAuthenticator(identityProvider configapi.Identit
...
+		connectionInfo := provider.RemoteConnectionInfo
+		if len(connectionInfo.URL) == 0 {
+			return nil, fmt.Errorf("URL is required for KeystonePasswordIdentityProvider")
+		}
+		transport, err := cmdutil.TransportFor(connectionInfo.CA, connectionInfo.ClientCert.CertFile, connectionInfo.ClientCert.KeyFile)
+		if err != nil {
+			return nil, fmt.Errorf("Error building KeystonePasswordIdentityProvider client: %v", err)
+		}
+
+		if !strings.HasPrefix(connectionInfo.URL, "https") {
```
Comment (liggitt) [inline pkg/cmd/server/origin/auth.go]: do we really want to log both here and in validation?

File: pkg/cmd/server/api/validation/oauth.go [hunk 8]
```diff
@@ -160,6 +164,18 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
...
+		return validationResults
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: any validation of DomainName needed?

Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: actually, are we requiring the v3 API? If so, is domainName required, and should we be using `AuthenticateV3()`?

File: pkg/cmd/server/api/validation/oauth.go (hunk 5 above)
Comment (sseago) [inline pkg/cmd/server/api/validation/oauth.go]: Fixing this

File: pkg/cmd/server/api/validation/oauth.go (hunk 6 above)
Comment (sseago) [inline pkg/cmd/server/api/validation/oauth.go]: ok

File: pkg/cmd/server/origin/auth.go (hunk 7 above)
Comment (sseago) [inline pkg/cmd/server/origin/auth.go]: Probably not. I'll pull this bit out.

File: pkg/cmd/server/api/validation/oauth.go (hunk 8 above)
Comment (sseago) [inline pkg/cmd/server/api/validation/oauth.go]: I don't see anything obvious listed in the keystone API docs around domain name formatting, but we should make sure it's present and probably validate the URL against v3 (and call AuthenticateV3).

Comment (sseago) [inline pkg/cmd/server/api/validation/oauth.go]: Making the explicit AuthenticateV3 call is the way to go. This forces v3 even if the config URL omits the version.

**Comment by sseago** [comment]
I just updated the PR incorporating feedback and adding a working unit test against a mocked test keystone server.

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+	opts := gophercloud.AuthOptions{
+		IdentityEndpoint: a.url,
+		Username:         username,
+		Password:         password,
+		DomainName:       a.domainName,
+	}
+
+	// Calling NewClient/Authenticate manually rather than simply calling AuthenticatedClient
+	// in order to pass in a transport object that supports SSL
+	client, err := openstack.NewClient(opts.IdentityEndpoint)
```
Comment (liggitt) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: check err and return before using `client`... it might be nil

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+	client.HTTPClient = *a.client
+	if err == nil {
```
Comment (liggitt) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: no need to check err here if you check above

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	if err != nil {
+		glog.Infof("Failed: Starting openstack authenticate client: %v", err)
+		if responseErr, ok := err.(*gophercloud.UnexpectedResponseCodeError); ok {
+			if responseErr.Actual == 401 {
+				return nil, false, nil
+			}
+		}
+		if err == tokens.ErrMissingPassword {
```
Comment (liggitt) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: is an empty password ever valid? if not, can we return immediately on empty passwords?

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+			return nil, false, nil
+		}
+		return nil, false, errors.New("Failed to authenticate")
```
Comment (liggitt) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: return the original error so we don't lose details? maybe add logging as well, not sure what level it should be at

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword_test.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
+package keystonepassword
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"testing"
+	"io/ioutil"
+	
+//	"github.com/rackspace/gophercloud"
```
Comment (liggitt) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword_test.go]: clean up comments

File: pkg/cmd/server/api/validation/oauth.go [hunk 14]
```diff
@@ -162,6 +166,25 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
 	return validationResults
 }
 
+// RemoteConnection fields validated separately -- this is for keystone-specific validation
+func ValidateKeystoneIdentityProvider(provider *api.KeystonePasswordIdentityProvider, identityProvider api.IdentityProvider) ValidationResults {
+	validationResults := ValidationResults{}
+	validationResults.AddErrors(ValidateRemoteConnectionInfo(provider.RemoteConnectionInfo)...)
+
+	providerURL, err := url.Parse(provider.RemoteConnectionInfo.URL)
+	if err != nil {
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: ValidateRemoteConnectionInfo already adds an error if the url doesn't parse... you can just check the scheme if `err == nil`

**Comment by liggitt** [comment]
[test]

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+	// Calling NewClient/Authenticate manually rather than simply calling AuthenticatedClient
+	// in order to pass in a transport object that supports SSL
+	client, err := openstack.NewClient(opts.IdentityEndpoint)
+	client.HTTPClient = *a.client
+	if err == nil {
+		err = openstack.AuthenticateV3(client, opts)
+	}
+
+	if err != nil {
+		glog.Infof("Failed: Starting openstack authenticate client: %v", err)
```
Comment (liggitt) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: higher verbosity, probably V(4)

**Comment by liggitt** [comment]
thanks, just a few final things and a clean test run

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go (hunk 11 above)
Comment (sseago) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: It's not valid. The gophercloud library errors out with ErrMissingPassword without even making the call to keystone, but yeah if we do the same check ourselves it might make more sense than passing it on and checking for the error return.

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go (hunk 12 above)
Comment (sseago) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: I already have a logging statement above (first line in the block, above the 401 check). I can return the original error -- it gets swallowed by the UI code to display a generic "something went wrong" sort of error to the user anyway. Actually, come to think of it we probably don't need to log the error if it's a 401 response -- that's just a failed login, and if we get any other error here (other than the missing password one that we should handle differently), perhaps it should be a higher level log, since any error other than a 401 or a missing password probably indicates a config problem or a problem with keystone. Should it be glog.Errorf? (or Warningf)?

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword_test.go (hunk 13 above)
Comment (sseago) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword_test.go]: Oops. I missed that one.

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go (hunk 10 above)
Comment (sseago) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: So I think all I need to do here is move the assignment of client.HTTPClient into the "if err == nil" block. If there's no error, we set HTTPClient and authenticate in the "err isn't nil" block. If there is an error, that's handled in the block below.

Comment (liggitt) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: convention is to return immediately on errors

```
foo, err := op()
if err != nil {
  return nil, err
}

bar, err := op()
if err != nil {
  return nil, err
}
```

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go (hunk 12 above)
Comment (liggitt) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: I need to sweep how logging works for auth in general... how individual authenticators log, and how the server handles returned errors. For now, log at high verbosity on login failure, and log an error if you get to this line. I'll revisit.

File: pkg/cmd/server/api/v1/types.go [hunk 16]
```diff
@@ -544,6 +544,15 @@ type LDAPAttributeMapping struct {
 	Email []string `json:"email"`
 }
 
+// FIXME: add this to types_test.go
```
Comment (liggitt) [inline pkg/cmd/server/api/v1/types.go]: add this

**Comment by sseago** [comment]
Next round of cleanup pushed to PR...

File: pkg/auth/authenticator/password/keystonepassword/keystonepassword.go [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+				return nil, false, nil
+			}
+		}
+		glog.Warningf("Failed: Calling openstack AuthenticateV3: %v", err)
+		return nil, false, err
+	}
+
+	identity := authapi.NewDefaultUserIdentityInfo(a.providerName, username)
+	user, err := a.identityMapper.UserFor(identity)
+	glog.V(4).Infof("Got userIdentityMapping: %#v", user)
```
Comment (liggitt) [inline pkg/auth/authenticator/password/keystonepassword/keystonepassword.go]: move this after the `err != nil` check

File: pkg/cmd/server/api/v1/types.go [hunk 18]
```diff
@@ -544,6 +544,14 @@ type LDAPAttributeMapping struct {
 	Email []string `json:"email"`
 }
 
+type KeystonePasswordIdentityProvider struct {
+	v1.TypeMeta `json:",inline"`
+	// RemoteConnectionInfo contains information about how to connect to the external basic auth server
```
Comment (liggitt) [inline pkg/cmd/server/api/v1/types.go]: s/external basic auth server/keystone server/

**Comment by liggitt** [comment]
godoc nit, LGTM

**Comment by liggitt** [comment]
tests failed with:
```
hack/verify-gofmt.sh
!!! gofmt needs to be run on the following files: 
./pkg/auth/authenticator/password/keystonepassword/keystonepassword_test.go
```

**Comment by sseago** [comment]
verify-gofmt succeeds now. I'd run it on the other new files but forgot to run it on the test file.

**Comment by liggitt** [comment]
[test]

File: pkg/cmd/server/api/validation/oauth.go [hunk 19]
```diff
@@ -162,6 +166,24 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
...
+func ValidateKeystoneIdentityProvider(provider *api.KeystonePasswordIdentityProvider, identityProvider api.IdentityProvider) ValidationResults {
+	validationResults := ValidationResults{}
+	validationResults.AddErrors(ValidateRemoteConnectionInfo(provider.RemoteConnectionInfo)...)
+
+	providerURL, err := url.Parse(provider.RemoteConnectionInfo.URL)
+	if err == nil {
+		if providerURL.Scheme != "https" {
+			validationResults.AddWarnings(fielderrors.NewFieldInvalid("url", provider.RemoteConnectionInfo.URL, "Auth URL should be secure and start with https"))
+		}
+	}
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: I forgot, `ValidateSecureURL` already exists... I think this whole bit can simplify to 
```
validationResults.AddWarnings(ValidateSecureURL("url", provider.RemoteConnectionInfo.URL)
```

Comment (sseago) [inline pkg/cmd/server/api/validation/oauth.go]: So I looked into this. What was needed was similar to what you have there:
```
       _, urlErrs := ValidateSecureURL(provider.RemoteConnectionInfo.URL, "url")
       validationResults.AddWarnings(urlErrs...)
```
The only issue is while it acts as a warning (and not a server-stopping error), the log message implies a real error:
```
W1014 15:16:28.928500   28181 start_master.go:278] oauthConfig.identityProvider[0].provider.url: invalid value 'http://192.168.0.198:5000/v3', Details: must use https scheme
```
With the custom validation we avoid the inaccurate message. I could refactor ValidateSecureURL to take a 'warning' vs. 'error' boolean arg, but that might be more confusing than just creating a WarnOnInsecureURL method, but with only one caller of it, I don't know if that's any better than what I already have above.

**Comment by sseago** [comment]
@liggitt  What's left to do here. Jenkins is succeeding but Travis failed. Is this a known issue with the travis results, or does this represent an actual problem with this PR that needs fixing?

**Comment by liggitt** [comment]
You're all set, thanks. I'll work on getting this merged tomorrow.

**Comment by sseago** [comment]
Excellent. Thanks for the update.

**Comment by liggitt** [comment]
[test]

**Comment by liggitt** [comment]
[test]

**Comment by liggitt** [comment]
[merge]
